### PR TITLE
[tune] Introduce TrialCheckpoint class, making checkpoint down/upload easie

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Currently there is no smooth way to access checkpoints on cloud storage after a run. This is especially true when e.g. the best checkpoint was trained on a worker node (and not the head node) - the checkpoint lives on cloud storage but may have to be downloaded to the driver node for postprocessing.

This PR adds a `TrialCheckpoint` abstraction that is returned by `ExperimentAnalysis.get_best_checkpoint()` and `ExperimentAnalysis.best_checkpoint` (instead of the local trial path previously, which did not ensure consistency or even local existence of the checkpoint).

The `TrialCheckpoint` keeps a record of a local and cloud location and offers utilities to sync between these:

- the `TrialCheckpoint.download()` method allows downloading from cloud storage to local storage
- the `TrialCheckpoint.upload()` method allows uploading from local storage to cloud storage
- the `TrialCheckpoint.save()` method allows saving to both local and cloud storage and fetches the checkpoint from cloud before if needed

This PR adds a new module, `ray.tune.cloud`, which includes the checkpoint syncing logic. This is somewhat of a duplicate to the `ray.tune.sync_client` abstractions. This is currently sensible in order to avoid refactoring sync clients in this PR. In a future PR, this should be consolidated.

The PR also adds API and local/CI end to end tests. In a follow-up PR, the features should be tested in the tune cloud checkpointing release tests.

Example usage:

```
    best_checkpoint = analysis.best_checkpoint
    print("Remote checkpoint path", best_checkpoint.cloud_path)
    best_checkpoint.save("./out_checkpoint_dir")
```

Backwards compatibility is achieved by making this a `os.PathLike` class.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/cloud.py
Comment (xwjiang2010) [inline python/ray/tune/cloud.py:69]: I really like this idea of pairing local path and cloud path. Ideally I think things should be organized as:
on cloud: cloud_path_prefix  + common_upload_path
local: local_path_prefix + common_upload_path

We can use this to simplify the logic across the board.

One thing I am not so sure about is  why do we need to keep track of `self.metrics` here? This seems to be overly coupled with ExperiementAnalysis and will make this TrialCheckpoint class less generic.

File: python/ray/tune/cloud.py
Comment (xwjiang2010) [inline python/ray/tune/cloud.py:91]: The overwrite option as well as the force download is a bit confusing to me as far as API is concerned.

Is that because a pair of local and cloud path won't cover it? Somehow we will change local path in the middle?

If possible, I would prefer to have a very scoped API and corresponding functionality.

File: python/ray/tune/cloud.py
Comment (krfricke) [reply python/ray/tune/cloud.py:69]: The initial idea for metric was to sometime extend this to a full `last_results` dict or so, so that `analysis.best_checkpoint.last_results` could be examined.

I'll get rid of that for now though, we can add this if there's demand for it.

Comment (krfricke) [reply python/ray/tune/cloud.py:69]: +1 to the prefix / common path abstraction. Let's go with the current replacement logic for now, but when refactoring the tech debt here we should do this!

File: python/ray/tune/cloud.py
Comment (krfricke) [reply python/ray/tune/cloud.py:91]: The reason why I'd like to keep `overwrite` is that I imagine a use case where a checkpoint to be served will be stored in the same local dir or cloud dir, e.g. something like `checkpoint.save("/home/ray/serve_checkpoint")`. If this location is updated with a newer checkpoint after training again, it is useful to be able to overwrite it from the API rather than having to delete it manually before.

Though after reviewing the code I'm not married to this API. I think the use case might be common enough to warrant the API, but happy to hear your and other people's thoughts on this

Comment (krfricke) [reply python/ray/tune/cloud.py:87]: Ah there is actually one CI test where we rely on this, but we could potentially change the scope.

File: python/ray/tune/cloud.py
Comment (Yard1) [inline python/ray/tune/cloud.py:24]: The ifs may get hard to maintain if we ever want to add support for other clouds. Maybe it would be a good idea to have a small class defining commands for a cloud, and then have a list/dict of them.

Comment (gjoliver) [reply python/ray/tune/cloud.py:24]: I 2nd this sentiment. ideally we would have all storage providers implement a common api class, so commands and edge cases are grouped per provider.
but I understand that is a much bigger change, and if you'd rather do it in a followup change.

File: python/ray/tune/cloud.py
Comment (gjoliver) [inline python/ray/tune/cloud.py:300]: there are a lot of if elses here, handling local vs cloud, etc cases. but if we make local storage also a "provider", you don't need to have any special if handling here.
supposedly if source is a local path, we can just use the local provider module to handle it, same as any other cloud provider. and this part of the logic will be much cleaner.

File: python/ray/tune/cloud.py
Comment (gjoliver) [reply python/ray/tune/cloud.py:90]: I guess overwrite is at least intuitive to understand, and seems useful.
update_local_path though, is a pretty confusing parameter.
why would we not want to update it? is it really necessary?

File: python/ray/tune/cloud.py
Comment (krfricke) [reply python/ray/tune/cloud.py:24]: We'll definitely do it (and presumably get rid of the command line interface and use proper SDKs instead), but it's out of scope for this PR.

File: python/ray/tune/cloud.py
Comment (krfricke) [reply python/ray/tune/cloud.py:109]: We usually won't want to update it if it points to a temporary directory. Otherwise we can default to updating it whenever it is not set. I'll see if I can simplify this in the code

Comment (krfricke) [reply python/ray/tune/cloud.py]: I removed this from the API - thanks for the suggestion

File: python/ray/tune/cloud.py
Comment (krfricke) [reply python/ray/tune/cloud.py:303]: A difference here is that we explicitly distinguish between a local and a cloud location in the object and would likely want to do so in terms of API.
However, I agree, once we have a Provider API, we can just do `download()` and then `copytree` the local directory.

Comment (krfricke) [reply python/ray/tune/cloud.py:303]: My proposal here would be, as long as everyone is fine with the API, to go with the current implementation to enable this case, and go to a more general storage solution later (presumably whilst refactoring the syncing logic generally)

Comment (gjoliver) [reply python/ray/tune/cloud.py:303]: happy to discuss why we need to differentiate and if we can hide the difference behind a common api.
but that's definitely for a separate change.
thanks.

**Review by gjoliver — APPROVED** [review]
Seems reasonable to me, but definitely wait for xiaowei's approval.

**Review by xwjiang2010 — APPROVED** [review]
Let's move forward to unblock GA first. Thanks Kai!

**Pull request merged**
